src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs

```diff
@@ -0,0 +1,311 @@
+                }
+
+                if (Directory.Exists(packagePath))
+                {
+                    throw new ToolPackageException(
+                        string.Format(
+                            CommonLocalizableStrings.ToolPackageConflictPackageId,
+                            packageId,
+                            version.ToNormalizedString()));
+                }
+
+                var packageHash = Convert.ToBase64String(new CryptoHashProvider("SHA512").CalculateHash(reader.GetNuspec()));
+                var hashPath = new VersionFolderPathResolver(hashPathLocation).GetHashPath(packageId.ToString(), version);
+
+                Directory.CreateDirectory(Path.GetDirectoryName(hashPath));
+                File.WriteAllText(hashPath, packageHash);
+            }
+
+            // Extract the package
+            var nupkgDir = Path.Combine(hashPathLocation, packageId.ToString(), version.ToString());
+            var filesInPackage = await _nugetPackageDownloader.ExtractPackageAsync(packagePath, new DirectoryPath(nupkgDir));
+
+            return version;
+        }
+
+        private static void CreateAssetFiles(
+            PackageId packageId,
+            NuGetVersion version,
+            DirectoryPath nugetLocalRepository,
+            DirectoryPath assetFileDirectory,
+            string runtimeJsonGraph)
+        {
+            // To get runtimeGraph:
+            var runtimeGraph = JsonRuntimeFormat.ReadRuntimeGraph(runtimeJsonGraph);
+
+            // Create ManagedCodeConventions:
+            var conventions = new ManagedCodeConventions(runtimeGraph);
+
+            //  Create LockFileTargetLibrary
+            var lockFileLib = new LockFileTargetLibrary()
+            {
+                Name = packageId.ToString(),
+                Version = version,
+                Type = LibraryType.Package,
+                PackageType = new List<PackageType>() { PackageType.DotnetTool }
+            };
+
+            //  Create NuGetv3LocalRepository
+            NuGetv3LocalRepository localRepository = new(nugetLocalRepository.Value);
+            var package = localRepository.FindPackage(packageId.ToString(), version);
+
+            var collection = new ContentItemCollection();
+            collection.Load(package.Files);
+
+            //  Create criteria
+            var managedCriteria = new List<SelectionCriteria>(1);
+            // Use major.minor version of currently running version of .NET
+            var currentTargetFramework = new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, new Version(Environment.Version.Major, Environment.Version.Minor));
+
+            var standardCriteria = conventions.Criteria.ForFrameworkAndRuntime(
+                currentTargetFramework,
+                RuntimeInformation.RuntimeIdentifier);
+            managedCriteria.Add(standardCriteria);
+
+            //  Create asset file
+            if (lockFileLib.PackageType.Contains(PackageType.DotnetTool))
```

Comment:
This is the only place where we check the `DotnetTool` type, how does it manifest when a package without the correct type is attempted to be installed?

Reply:
I think it'd be great to add a test for this. 

Currently all of this validation happens NuGet side and that's where the tests are.

We have https://github.com/NuGet/NuGet.Client/blob/40648c3cb3de37cf62b194accdc8e7117b36245f/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetToolTests.cs I think it'd be g reat to just make sure each of the appropriate tests there has an equivalent in this repo.

Reply:
Thank you for the comment~To be follow up in 35289